File: 1917-maximum-average-pass-ratio/1917-maximum-average-pass-ratio.py
class Solution:
    def maxAverageRatio(self, classes: List[List[int]], extraStudents: int) -> float:
        def calculate_Increase_In_Pass_Ratio(passes,total):
            return ((passes+1)/(total+1))-(passes/total)
        heap=[]
        for p,t in classes:
            heapq.heappush(heap,(-calculate_Increase_In_Pass_Ratio(p,t),p,t))
        print(list(heap))
        for i in range(extraStudents):
            _,p,t=heapq.heappop(heap)
            heapq.heappush(heap,(-calculate_Increase_In_Pass_Ratio(p+1,t+1),p+1,t+1))
        avg=0
        print(list(heap))
        while heap:
            _,p,t=heapq.heappop(heap)
            avg+=p/t
        return avg/len(classes)

File: 1917-maximum-average-pass-ratio/1917-maximum-average-pass-ratio.java
class Solution {
    public double calculateIncreaseInPassRatio(double pass,double total){
        return ((double)(pass+1)/(total+1))-((double)pass/total);
    }
    public double maxAverageRatio(int[][] classes, int extraStudents) {
        PriorityQueue<double[]> heap=new PriorityQueue<>((a,b)->Double.compare(b[0],a[0]));
        for(int[] cl:classes){
            double pass=cl[0],total=cl[1];
            heap.offer(new double[]{calculateIncreaseInPassRatio(pass,total),pass,total});
        }
        for(int i=0;i<extraStudents;i++){
            double[] top=heap.poll();
            double pass=top[1],total=top[2];
            heap.offer(new double[]{calculateIncreaseInPassRatio(pass+1,total+1),pass+1,total+1});
        }
        double avg=0;
        while(!heap.isEmpty()){
            double[] top=heap.poll();
            double pass=top[1],total=top[2];
            avg+=((double)pass)/total;
        }
        return avg/classes.length;
    }
}